### DISCUSSION TITLE
How to add scrollmagic to rollup.config.js?

### DISCUSSION THREAD

**Discussion by gurds91** [holger1411/understrap#1811] [Q&A]
Hi guys,

I used npm install scrollmagic command in terminal to install scroll magic. js 

I believe it needs to be added to rollup.config.js , how to do this- what do I need to type in 

the relative path is node_modules\scrollmagic\scrollmagic\uncompressed\ScrollMagic.js 

Also do you need to enqueue this script into functions.php as well? Code to do this as well would be great. 

After that then you can just input your code into custom-javscript.js as normal to make use of scrollmagic as well.

**Comment by IanDelMar** [comment]
> Also do you need to enqueue this script into functions.php as well? Code to do this as well would be great.

It depends. You can add the script to the bundle using rollup.config.js. Or you can load it as an individual file by enqueuing it.

For bundling the script adjust your rollup.config.js like so 
```diff
module.exports = {
	input: [
		path.resolve( __dirname, '../js/bootstrap.js' ),
		path.resolve( __dirname, '../js/skip-link-focus-fix.js' ),
+		path.resolve( __dirname, '../../node_modules/scrollmagic/scrollmagic/uncompressed/ScrollMagic.js' ),
		path.resolve( __dirname, '../js/custom-javascript.js' ),
	],
	...
```

For loading the file individually you can use a script to copy ScrollMagic.min.js to your js folder or you can let the rollup plugin [rollup-plugin-copy](https://www.npmjs.com/package/rollup-plugin-copy) doing it for you.

For using the rollup plugin type `npm install rollup-plugin-copy --save-dev` and adjust your rollup.config.js like so
```diff
const path = require('path')
const { babel } = require('@rollup/plugin-babel')
const { nodeResolve } = require('@rollup/plugin-node-resolve')
import commonjs from '@rollup/plugin-commonjs'
import multi from '@rollup/plugin-multi-entry'
const replace = require('@rollup/plugin-replace')
const banner = require('./banner.js')
+ const copy = require('rollup-plugin-copy')

...

const plugins = [
  babel({
    // Only transpile our source code
    exclude: 'node_modules/**',
    // Include the helpers in the bundle, at most one copy of each
    babelHelpers: 'bundled'
  }),
  replace({
      'process.env.NODE_ENV': '"production"',
      preventAssignment: true
  }),
  nodeResolve(),
  commonjs(),
  multi(),
+  copy( {
+    targets: [ {
+      src: [
+        './node_modules/scrollmagic/scrollmagic/uncompressed/ScrollMagic.js',
+        './node_modules/scrollmagic/scrollmagic/minified/ScrollMagic.min.js'
+      ],
+      dest: './js'
+  } ),
]

...
```
to enqueue the magic scroll script file put this in your functions.php
```php
function slug_enqueue_scroll_magic() {
  $file    = '/js/ScrollMagic.min.js';
  $version = filemtime( get_template_directory() . $file );

  wp_enqueue_script( 'scroll-magic', get_template_directory_uri() . $file, array(), $version, true );
}
add_action( 'wp_enqueue_scripts', 'slug_enqueue_scroll_magic' );
```

Then put your script that uses scroll magic into custom-javascript.js and make sure ScrollMagic.min.js is loaded before 'understrap-scripts'.

**Reply by gurds91** [reply]
I see so the first option adding to rollup.config.js is easier?

**Reply by IanDelMar** [reply]
Yes. But in an HTTP2 world browsers are able to download assets in parallel so there might be a performance benefit of loading the files separately. Having separate files also allows to load a script on specific pages only.

**Reply by gurds91** [reply]
by putting in scroll magic into inputs array before the custom javascript.js does this mean that scroll magic will run before the custom javascript js?

**Reply by IanDelMar** [reply]
Revision of my original answer. Do not add Scroll Magic in your rollup.config.js. I was focused on looking for an esm module and not finding one didn't realize that you can import Scroll Magic.

Directly import it e.g. in your custom-javascript.js like so:

```js
// custom-javascript.js
import ScrollMagic from "scrollmagic/scrollmagic/uncompressed/ScrollMagic";

const controller = new ScrollMagic.Controller();
...
```
Sorry for the confusion.
